meta-mender-core/recipes-mender/mender/mender.inc

```diff
@@ -122,9 +122,28 @@ python do_prepare_mender_conf() {
     # Tenant-token is optional, but falls back to a default-value set in config.go
     conf_maybe_add("TenantToken", "MENDER_TENANT_TOKEN", getvar=True, integer=False)
 
-    dst_conf = os.path.join(d.getVar("B"), "mender.conf")
+    # Split the mender.conf file into a transient and a persistent configuration file.
+    persistent_configs = ["RootfsPartA", "RootfsPartB"]
+
+    persistent_conf = {}
+
+    # Extract the variables that are destined for the persistent mender-configuration.
+    for config_var in transient_conf:
+        if config_var in persistent_configs:
+          persistent_conf[config_var] = transient_conf[config_var]
```

Comment:
It just occurred to me that this is not upgrade-safe. Imagine if you have an existing device that has everything in `/etc/mender/mender.conf`. Its new update will have stuff split into `/etc` and `/data`, but `/data` is not distributed in updates, only in initial deployments. Hence the device will be missing this information once it reboots into the new update. And there is no simple workaround with the current approach.

I think we can solve this by introducing a `PACKAGECONFIG` option, say `split-mender-conf`, which defaults to on, but which can be turned off for those who need it. Essentially it would be along these lines:

```bash
# Just needs to defined. It normally deals with dependencies,
# but we don't need any here.
PACKAGECONFIG[split-mender-conf] = ",,,"

PACKAGECONFIG_append = " split-mender-conf"

# And then in every place where you need to make a decision based on it.
# Python:
bb.utils.contains('PACKAGECONFIG', 'split-mender-conf', True, False, d)
# Shell
${bb.utils.contains('PACKAGECONFIG', 'split-mender-conf', 'true', 'false', d)}
```

So then you simply don't do any splitting if it's not set. What do you think?

Reply:
This should also be mentioned in the changelog, what has happened, and how to turn it off for those who need it. And maybe even an entry in our troubleshooting guide.

Reply:
Perhaps @mirzak or @drewmoseley have some input also. Basically, how do we ensure that upgrading works if we start splitting `mender.conf` across `/etc` and `/data`.

Reply:
Is it possible to create the /data/ contents with a state script of some kind?

Reply:
That's also a possibility. One advantage of doing it that way is that we would properly migrate everyone, instead of keeping support for the old way indefinitely. We could make this as an optional script, and document the upgrade procedure.

Reply:
I feel like it should be the other way around. You have to enable this functionality, rather than disable it. Thus migrating will be an option, but strongly advised. However, I do not have any strong feelings regarding this. First taker gets the decision.

Reply:
From a compatibility perspective, I see your point, but I think this is too detailed that most people will care enough to enable it, so I still think that our recommendation (which is to split) should be the default.

It will of course mean that we cannot backport this, but this is rather expected I would say.